File: js/src/backends/registry.js
/**
 * Backend registry for links-queue.
 *
 * This module provides a registry for storage backend implementations,
 * allowing backends to be registered and instantiated by name via configuration.
 *
 * @module backends/registry
 *
 * @example
 * // Register a custom backend
 * BackendRegistry.register('my-custom', MyCustomBackend);
 *
 * // Create backend via configuration
 * const backend = BackendRegistry.create({
 *   type: 'my-custom',
 *   options: { /* custom options *\/ }
 * });
 */

import { MemoryLinkStore } from './memory.js';

/**
 * A wrapper that adapts MemoryLinkStore to the StorageBackend interface.
 *
 * This allows MemoryLinkStore to be used as a storage backend while maintaining
 * backward compatibility with its existing interface.
 *
 * @implements {import('./types.ts').StorageBackend}
 */
class MemoryBackendAdapter {
  /**
   * Creates a new MemoryBackendAdapter.
   *
   * @param {import('./types.ts').MemoryBackendOptions} [options] - Configuration options
   */
  constructor(options = {}) {
    /** @private */
    this._store = new MemoryLinkStore();

    /** @private */
    this._options = options;

    /** @private */
    this._connected = false;

    /** @private */
    this._connectedAt = null;

    /** @private */
    this._stats = {
      reads: 0,
      writes: 0,
      deletes: 0,
      queries: 0,
    };
  }

  /**
   * Establishes connection to the backend.
   * For memory backend, this is essentially a no-op but tracks connection state.
   *
   * @returns {Promise<void>}
   */
  // eslint-disable-next-line require-await
  async connect() {
    if (this._connected) {
      return;
    }
    this._connected = true;
    this._connectedAt = new Date().toISOString();
  }

  /**
   * Disconnects from the backend.
   * For memory backend, this clears the connection state.
   *
   * @returns {Promise<void>}
   */
  // eslint-disable-next-line require-await
  async disconnect() {
    this._connected = false;
    this._connectedAt = null;
  }

  /**
   * Checks if the backend is connected.
   *
   * @returns {boolean}
   */
  isConnected() {
    return this._connected;
  }

  /**
   * Saves a link to storage.
   *
   * @param {import('../types.js').Link} link - The link to save
   * @returns {Promise<import('../types.js').LinkId>}
   */
  async save(link) {
    this._ensureConnected();
    this._stats.writes++;

    // If the link has an ID of 0 or is "nothing", create a new link
    if (link.id === 0 || link.id === '' || link.id === 0n) {
      const created = link.values
        ? await this._store.createWithValues(
            link.source,
            link.target,
            link.values
          )
        : await this._store.create(link.source, link.target);
      return created.id;
    }

    // Otherwise, update existing or create with the given ID
    // For memory backend, we'll try to update first, then create if not exists
    const existing = await this._store.get(link.id);
    if (existing) {
      const updated = await this._store.update(
        link.id,
        link.source,
        link.target
      );
      return updated.id;
    }

    // Create new link (will use auto-generated ID)
    const created = link.values
      ? await this._store.createWithValues(
          link.source,
          link.target,
          link.values
        )
      : await this._store.create(link.source, link.target);
    return created.id;
  }

  /**
   * Loads a link by ID.
   *
   * @param {import('../types.js').LinkId} id - The link ID
   * @returns {Promise<import('../types.js').Link | null>}
   */
  // eslint-disable-next-line require-await
  async load(id) {
    this._ensureConnected();
    this._stats.reads++;
    return this._store.get(id);
  }

  /**
   * Deletes a link by ID.
   *
   * @param {import('../types.js').LinkId} id - The link ID
   * @returns {Promise<boolean>}
   */
  // eslint-disable-next-line require-await
  async delete(id) {
    this._ensureConnected();
    this._stats.deletes++;
    return this._store.delete(id);
  }

  /**
   * Queries links by pattern.
   *
   * @param {import('../types.js').LinkPattern} pattern - The pattern to match
   * @returns {Promise<import('../types.js').Link[]>}
   */
  // eslint-disable-next-line require-await
  async query(pattern) {
    this._ensureConnected();
    this._stats.queries++;
    return this._store.find(pattern);
  }

  /**
   * Saves multiple links in batch.
   *
   * @param {readonly import('../types.js').Link[]} links - Links to save
   * @returns {Promise<import('../types.js').LinkId[]>}
   */
  async saveBatch(links) {
    this._ensureConnected();
    const ids = [];
    for (const link of links) {
      const id = await this.save(link);
      ids.push(id);
    }
    return ids;
  }

  /**
   * Deletes multiple links in batch.
   *
   * @param {readonly import('../types.js').LinkId[]} ids - IDs to delete
   * @returns {Promise<boolean[]>}
   */
  async deleteBatch(ids) {
    this._ensureConnected();
    const results = [];
    for (const id of ids) {
      const result = await this.delete(id);
      results.push(result);
    }
    return results;
  }

  /**
   * Returns backend capabilities.
   *
   * @returns {import('./types.ts').BackendCapabilities}
   */
  getCapabilities() {
    return {
      supportsTransactions: false,
      supportsBatchOperations: false, // We simulate batch ops
      durabilityLevel: 'none',
      maxLinkSize: 0, // Unlimited
      supportsPatternQueries: true,
    };
  }

  /**
   * Returns backend statistics.
   *
   * @returns {import('./types.ts').BackendStats}
   */
  getStats() {
    const now = Date.now();
    const connectedTime = this._connectedAt
      ? new Date(this._connectedAt).getTime()
      : now;

    return {
      totalLinks: this._store._links.size,
      usedSpace: 0, // Not tracked for memory backend
      operations: { ...this._stats },
      connectedAt: this._connectedAt,
      uptimeMs: this._connected ? now - connectedTime : 0,
    };
  }

  /**
   * Ensures the backend is connected before operations.
   *
   * @private
   * @throws {Error} If not connected
   */
  _ensureConnected() {
    if (!this._connected) {
      throw new Error('Backend is not connected. Call connect() first.');
    }
  }

  /**
   * Clears all data (for testing).
   *
   * @returns {Promise<void>}
   */
  async clear() {
    await this._store.clear();
    this._stats = {
      reads: 0,
      writes: 0,
      deletes: 0,
      queries: 0,
    };
  }
}

/**
 * Registry for storage backend implementations.
 *
 * Provides methods to register custom backends and create backend instances
 * from configuration.
 *
 * @example
 * // Register a custom backend
 * BackendRegistry.register('redis', RedisBackend);
 *
 * // Create from config
 * const backend = BackendRegistry.create({ type: 'redis', options: { host: 'localhost' } });
 */
class BackendRegistryClass {
  constructor() {
    /**
     * Map of backend type names to constructors/factories.
     * @private
     * @type {Map<string, import('./types.ts').BackendConstructor | import('./types.ts').BackendFactory>}
     */
    this._backends = new Map();

    // Register built-in backends
    this._backends.set('memory', MemoryBackendAdapter);
  }

  /**
   * Registers a backend implementation.
   *
   * @param {string} name - The name to register the backend under
   * @param {import('./types.ts').BackendConstructor | import('./types.ts').BackendFactory} backend - The backend constructor or factory
   * @throws {Error} If name is empty or backend is invalid
   *
   * @example
   * // Register with constructor
   * BackendRegistry.register('my-backend', MyBackend);
   *
   * // Register with factory
   * BackendRegistry.register('complex-backend', (opts) => new ComplexBackend(opts));
   */
  register(name, backend) {
    if (!name || typeof name !== 'string') {
      throw new Error('Backend name must be a non-empty string');
    }

    if (typeof backend !== 'function') {
      throw new Error('Backend must be a constructor or factory function');
    }

    this._backends.set(name, backend);
  }

  /**
   * Unregisters a backend implementation.
   *
   * @param {string} name - The name of the backend to unregister
   * @returns {boolean} True if the backend was unregistered, false if it wasn't registered
   *
   * @example
   * BackendRegistry.unregister('my-backend');
   */
  unregister(name) {
    return this._backends.delete(name);
  }

  /**
   * Checks if a backend is registered.
   *
   * @param {string} name - The backend name to check
   * @returns {boolean} True if registered
   *
   * @example
   * if (BackendRegistry.has('redis')) {
   *   const backend = BackendRegistry.create({ type: 'redis' });
   * }
   */
  has(name) {
    return this._backends.has(name);
  }

  /**
   * Creates a backend instance from configuration.
   *
   * @param {import('./types.ts').BackendOptions} config - Backend configuration
   * @returns {import('./types.ts').StorageBackend} The created backend instance
   * @throws {Error} If the backend type is not registered
   *
   * @example
   * const memoryBackend = BackendRegistry.create({ type: 'memory' });
   *
   * const customBackend = BackendRegistry.create({
   *   type: 'my-custom',
   *   options: { host: 'localhost', port: 6379 }
   * });
   */
  create(config) {
    const { type, options } = config;

    const Backend = this._backends.get(type);
    if (!Backend) {
      throw new Error(
        `Unknown backend type: "${type}". ` +
          `Available types: ${this.listTypes().join(', ')}`
      );
    }

    // Try to instantiate - could be constructor or factory
    try {
      // Check if it's a class (has prototype with constructor)
      if (Backend.prototype && Backend.prototype.constructor === Backend) {
        return new Backend(options);
      }
      // Otherwise it's a factory function
      return Backend(options);
    } catch (error) {
      throw new Error(`Failed to create backend "${type}": ${error.message}`);
    }
  }

  /**
   * Lists all registered backend types.
   *
   * @returns {string[]} Array of registered backend type names
   *
   * @example
   * const types = BackendRegistry.listTypes();
   * console.log(types); // ['memory', 'my-custom', ...]
   */
  listTypes() {
    return Array.from(this._backends.keys());
  }

  /**
   * Resets the registry to default state (only built-in backends).
   *
   * Primarily useful for testing.
   */
  reset() {
    this._backends.clear();
    this._backends.set('memory', MemoryBackendAdapter);
  }
}

/**
 * Singleton instance of the backend registry.
 *
 * @type {BackendRegistryClass}
 *
 * @example
 * import { BackendRegistry } from 'links-queue';
 *
 * // Register custom backend
 * BackendRegistry.register('redis', RedisBackend);
 *
 * // Create backend
 * const backend = BackendRegistry.create({ type: 'redis' });
 */
export const BackendRegistry = new BackendRegistryClass();

/**
 * The memory backend adapter class (exported for direct use).
 */
export { MemoryBackendAdapter };

File: rust/src/backends/registry.rs
//! Backend registry for links-queue.
//!
//! This module provides a registry for storage backend implementations,
//! allowing backends to be registered and instantiated by name via configuration.
//!
//! # Design
//!
//! The registry uses type-erased factory functions to allow dynamic backend
//! creation without requiring trait objects for the backends themselves.
//!
//! # Example
//!
//! ```rust
//! use links_queue::{BackendRegistry, MemoryBackend, BackendConfig};
//!
//! // Create a registry
//! let mut registry = BackendRegistry::<u64>::new();
//!
//! // Built-in "memory" backend is pre-registered
//! assert!(registry.has("memory"));
//!
//! // Create a backend from config
//! let backend = registry.create(&BackendConfig::memory()).unwrap();
//! ```

use std::collections::HashMap;
use std::sync::Arc;

use crate::LinkType;

use super::memory_backend::MemoryBackend;
use super::traits::{BackendError, BackendResult, StorageBackend};

// =============================================================================
// Backend Configuration
// =============================================================================

/// Configuration for backend creation.
///
/// This struct holds the backend type name and any type-specific options
/// needed to create a backend instance.
///
/// # Example
///
/// ```rust
/// use links_queue::BackendConfig;
///
/// // Memory backend (default)
/// let config = BackendConfig::memory();
///
/// // Memory backend with capacity hint
/// let config = BackendConfig::memory_with_capacity(10000);
///
/// // Custom backend
/// let config = BackendConfig::new("my-custom")
///     .with_option("host", "localhost")
///     .with_option("port", "6379");
/// ```
#[derive(Debug, Clone)]
pub struct BackendConfig {
    /// The backend type name.
    pub backend_type: String,

    /// Type-specific options as key-value pairs.
    pub options: HashMap<String, String>,
}

impl BackendConfig {
    /// Creates a new backend configuration.
    ///
    /// # Arguments
    ///
    /// * `backend_type` - The name of the backend type
    ///
    /// # Example
    ///
    /// ```rust
    /// use links_queue::BackendConfig;
    ///
    /// let config = BackendConfig::new("redis");
    /// ```
    #[must_use]
    pub fn new(backend_type: impl Into<String>) -> Self {
        Self {
            backend_type: backend_type.into(),
            options: HashMap::new(),
        }
    }

    /// Creates a configuration for the memory backend.
    ///
    /// # Example
    ///
    /// ```rust
    /// use links_queue::BackendConfig;
    ///
    /// let config = BackendConfig::memory();
    /// ```
    #[must_use]
    pub fn memory() -> Self {
        Self::new("memory")
    }

    /// Creates a configuration for the memory backend with a capacity hint.
    ///
    /// # Arguments
    ///
    /// * `capacity` - Initial capacity for the link storage
    ///
    /// # Example
    ///
    /// ```rust
    /// use links_queue::BackendConfig;
    ///
    /// let config = BackendConfig::memory_with_capacity(10000);
    /// ```
    #[must_use]
    pub fn memory_with_capacity(capacity: usize) -> Self {
        Self::new("memory").with_option("capacity", capacity.to_string())
    }

    /// Adds an option to the configuration (builder pattern).
    ///
    /// # Arguments
    ///
    /// * `key` - Option name
    /// * `value` - Option value
    ///
    /// # Example
    ///
    /// ```rust
    /// use links_queue::BackendConfig;
    ///
    /// let config = BackendConfig::new("redis")
    ///     .with_option("host", "localhost")
    ///     .with_option("port", "6379");
    /// ```
    #[must_use]
    pub fn with_option(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.options.insert(key.into(), value.into());
        self
    }

    /// Gets an option value.
    ///
    /// # Arguments
    ///
    /// * `key` - Option name
    ///
    /// # Returns
    ///
    /// The option value if present.
    #[must_use]
    pub fn get_option(&self, key: &str) -> Option<&str> {
        self.options.get(key).map(String::as_str)
    }
}

impl Default for BackendConfig {
    fn default() -> Self {
        Self::memory()
    }
}

// =============================================================================
// Backend Factory
// =============================================================================

/// Type alias for backend factory functions.
///
/// Factory functions create a boxed backend from a configuration.
type BackendFactory<T> =
    Arc<dyn Fn(&BackendConfig) -> BackendResult<T, Box<dyn StorageBackendDyn<T>>> + Send + Sync>;

/// Object-safe version of `StorageBackend` for use with registry.
///
/// This trait allows storing backends as trait objects while maintaining
/// the async interface through boxed futures.
#[allow(clippy::type_complexity)]
pub trait StorageBackendDyn<T: LinkType>: Send + Sync {
    /// Connects to the backend.
    fn connect_dyn(
        &mut self,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = BackendResult<T, ()>> + Send + '_>>;

    /// Disconnects from the backend.
    fn disconnect_dyn(
        &mut self,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = BackendResult<T, ()>> + Send + '_>>;

    /// Checks if connected.
    fn is_connected_dyn(&self) -> bool;

    /// Saves a link.
    fn save_dyn(
        &mut self,
        link: crate::Link<T>,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = BackendResult<T, T>> + Send + '_>>;

    /// Loads a link.
    fn load_dyn(
        &self,
        id: T,
    ) -> std::pin::Pin<
        Box<dyn std::future::Future<Output = BackendResult<T, Option<crate::Link<T>>>> + Send + '_>,
    >;

    /// Deletes a link.
    fn delete_dyn(
        &mut self,
        id: T,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = BackendResult<T, bool>> + Send + '_>>;

    /// Queries links.
    fn query_dyn<'a>(
        &'a self,
        pattern: &'a crate::LinkPattern<T>,
    ) -> std::pin::Pin<
        Box<dyn std::future::Future<Output = BackendResult<T, Vec<crate::Link<T>>>> + Send + 'a>,
    >;

    /// Returns capabilities.
    fn capabilities_dyn(&self) -> super::traits::BackendCapabilities;

    /// Returns stats.
    fn stats_dyn(&self) -> super::traits::BackendStats;
}

impl<T: LinkType, B: StorageBackend<T> + 'static> StorageBackendDyn<T> for B {
    fn connect_dyn(
        &mut self,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = BackendResult<T, ()>> + Send + '_>>
    {
        Box::pin(self.connect())
    }

    fn disconnect_dyn(
        &mut self,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = BackendResult<T, ()>> + Send + '_>>
    {
        Box::pin(self.disconnect())
    }

    fn is_connected_dyn(&self) -> bool {
        self.is_connected()
    }

    fn save_dyn(
        &mut self,
        link: crate::Link<T>,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = BackendResult<T, T>> + Send + '_>> {
        Box::pin(self.save(link))
    }

    fn load_dyn(
        &self,
        id: T,
    ) -> std::pin::Pin<
        Box<dyn std::future::Future<Output = BackendResult<T, Option<crate::Link<T>>>> + Send + '_>,
    > {
        Box::pin(self.load(id))
    }

    fn delete_dyn(
        &mut self,
        id: T,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = BackendResult<T, bool>> + Send + '_>>
    {
        Box::pin(self.delete(id))
    }

    fn query_dyn<'a>(
        &'a self,
        pattern: &'a crate::LinkPattern<T>,
    ) -> std::pin::Pin<
        Box<dyn std::future::Future<Output = BackendResult<T, Vec<crate::Link<T>>>> + Send + 'a>,
    > {
        Box::pin(self.query(pattern))
    }

    fn capabilities_dyn(&self) -> super::traits::BackendCapabilities {
        self.capabilities()
    }

    fn stats_dyn(&self) -> super::traits::BackendStats {
        self.stats()
    }
}

// =============================================================================
// Backend Registry
// =============================================================================

/// Registry for storage backend implementations.
///
/// The registry allows backends to be registered by name and created
/// from configuration. Built-in backends (like `memory`) are pre-registered.
///
/// # Type Parameters
///
/// * `T` - The link ID type (must implement [`LinkType`])
///
/// # Example
///
/// ```rust
/// use links_queue::{BackendRegistry, BackendConfig, MemoryBackend};
///
/// let mut registry = BackendRegistry::<u64>::new();
///
/// // Memory backend is pre-registered
/// let backend = registry.create(&BackendConfig::memory()).unwrap();
///
/// // List available backends
/// for name in registry.list_types() {
///     println!("Available: {}", name);
/// }
/// ```
pub struct BackendRegistry<T: LinkType> {
    /// Map of backend names to factory functions.
    factories: HashMap<String, BackendFactory<T>>,
}

impl<T: LinkType + 'static> Default for BackendRegistry<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T: LinkType + 'static> BackendRegistry<T> {
    /// Creates a new backend registry with built-in backends registered.
    ///
    /// # Example
    ///
    /// ```rust
    /// use links_queue::BackendRegistry;
    ///
    /// let registry = BackendRegistry::<u64>::new();
    /// assert!(registry.has("memory"));
    /// ```
    #[must_use]
    pub fn new() -> Self {
        let mut registry = Self {
            factories: HashMap::new(),
        };

        // Register built-in memory backend
        registry.register_memory_backend();

        registry
    }

    /// Registers the built-in memory backend.
    fn register_memory_backend(&mut self) {
        self.factories.insert(
            "memory".to_string(),
            Arc::new(|config: &BackendConfig| {
                let capacity = config
                    .get_option("capacity")
                    .and_then(|s| s.parse::<usize>().ok());

                let backend =
                    capacity.map_or_else(MemoryBackend::new, MemoryBackend::with_capacity);

                Ok(Box::new(backend) as Box<dyn StorageBackendDyn<T>>)
            }),
        );
    }

    /// Registers a custom backend factory.
    ///
    /// # Arguments
    ///
    /// * `name` - The name to register the backend under
    /// * `factory` - A function that creates backend instances from config
    ///
    /// # Example
    ///
    /// ```rust,ignore
    /// use links_queue::{BackendRegistry, BackendConfig, MemoryBackend};
    /// use std::sync::Arc;
    ///
    /// let mut registry = BackendRegistry::<u64>::new();
    ///
    /// // Register a custom backend factory
    /// registry.register("custom", Arc::new(|config: &BackendConfig| {
    ///     Ok(Box::new(MemoryBackend::<u64>::new()) as _)
    /// }));
    /// ```
    pub fn register<F>(&mut self, name: impl Into<String>, factory: Arc<F>)
    where
        F: Fn(&BackendConfig) -> BackendResult<T, Box<dyn StorageBackendDyn<T>>>
            + Send
            + Sync
            + 'static,
    {
        self.factories.insert(name.into(), factory);
    }

    /// Unregisters a backend.
    ///
    /// # Arguments
    ///
    /// * `name` - The name of the backend to unregister
    ///
    /// # Returns
    ///
    /// True if the backend was unregistered, false if it wasn't registered.
    pub fn unregister(&mut self, name: &str) -> bool {
        self.factories.remove(name).is_some()
    }

    /// Checks if a backend is registered.
    ///
    /// # Arguments
    ///
    /// * `name` - The backend name to check
    ///
    /// # Returns
    ///
    /// True if registered.
    #[must_use]
    pub fn has(&self, name: &str) -> bool {
        self.factories.contains_key(name)
    }

    /// Creates a backend instance from configuration.
    ///
    /// # Arguments
    ///
    /// * `config` - Backend configuration
    ///
    /// # Returns
    ///
    /// A boxed backend instance.
    ///
    /// # Errors
    ///
    /// Returns an error if the backend type is not registered or creation fails.
    ///
    /// # Example
    ///
    /// ```rust
    /// use links_queue::{BackendRegistry, BackendConfig};
    ///
    /// let registry = BackendRegistry::<u64>::new();
    /// let backend = registry.create(&BackendConfig::memory()).unwrap();
    /// ```
    pub fn create(
        &self,
        config: &BackendConfig,
    ) -> BackendResult<T, Box<dyn StorageBackendDyn<T>>> {
        let factory = self.factories.get(&config.backend_type).ok_or_else(|| {
            BackendError::Other(format!(
                "Unknown backend type: \"{}\". Available types: {}",
                config.backend_type,
                self.list_types().join(", ")
            ))
        })?;

        factory(config)
    }

    /// Lists all registered backend types.
    ///
    /// # Returns
    ///
    /// Vector of registered backend type names.
    #[must_use]
    pub fn list_types(&self) -> Vec<String> {
        self.factories.keys().cloned().collect()
    }

    /// Resets the registry to default state (only built-in backends).
    pub fn reset(&mut self) {
        self.factories.clear();
        self.register_memory_backend();
    }
}

// =============================================================================
// Tests
// =============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_backend_config_new() {
        let config = BackendConfig::new("test");
        assert_eq!(config.backend_type, "test");
        assert!(config.options.is_empty());
    }

    #[test]
    fn test_backend_config_memory() {
        let config = BackendConfig::memory();
        assert_eq!(config.backend_type, "memory");
    }

    #[test]
    fn test_backend_config_with_capacity() {
        let config = BackendConfig::memory_with_capacity(1000);
        assert_eq!(config.backend_type, "memory");
        assert_eq!(config.get_option("capacity"), Some("1000"));
    }

    #[test]
    fn test_backend_config_with_option() {
        let config = BackendConfig::new("redis")
            .with_option("host", "localhost")
            .with_option("port", "6379");

        assert_eq!(config.get_option("host"), Some("localhost"));
        assert_eq!(config.get_option("port"), Some("6379"));
        assert_eq!(config.get_option("nonexistent"), None);
    }

    #[test]
    fn test_registry_new() {
        let registry = BackendRegistry::<u64>::new();
        assert!(registry.has("memory"));
    }

    #[test]
    fn test_registry_create_memory() {
        let registry = BackendRegistry::<u64>::new();
        let result = registry.create(&BackendConfig::memory());
        assert!(result.is_ok());
    }

    #[test]
    fn test_registry_create_unknown() {
        let registry = BackendRegistry::<u64>::new();
        let result = registry.create(&BackendConfig::new("nonexistent"));
        assert!(result.is_err());
    }

    #[test]
    fn test_registry_list_types() {
        let registry = BackendRegistry::<u64>::new();
        let types = registry.list_types();
        assert!(types.contains(&"memory".to_string()));
    }

    #[test]
    fn test_registry_unregister() {
        let mut registry = BackendRegistry::<u64>::new();
        assert!(registry.has("memory"));
        assert!(registry.unregister("memory"));
        assert!(!registry.has("memory"));
        assert!(!registry.unregister("memory")); // Already unregistered
    }

    #[test]
    fn test_registry_reset() {
        let mut registry = BackendRegistry::<u64>::new();
        registry.unregister("memory");
        assert!(!registry.has("memory"));

        registry.reset();
        assert!(registry.has("memory"));
    }

    #[tokio::test]
    async fn test_created_backend_works() {
        let registry = BackendRegistry::<u64>::new();
        let mut backend = registry.create(&BackendConfig::memory()).unwrap();

        backend.connect_dyn().await.unwrap();
        assert!(backend.is_connected_dyn());

        let link = crate::Link::new(0, crate::LinkRef::Id(1), crate::LinkRef::Id(2));
        let id = backend.save_dyn(link).await.unwrap();
        assert_ne!(id, 0);

        let loaded = backend.load_dyn(id).await.unwrap();
        assert!(loaded.is_some());

        backend.disconnect_dyn().await.unwrap();
    }
}
